src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBatchMultiSource.java

```diff
@@ -75,26 +77,50 @@ public SalesforceBatchMultiSource(SalesforceMultiSourceConfig config) {
   @Override
   public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
     StageConfigurer stageConfigurer = pipelineConfigurer.getStageConfigurer();
-    config.validate(stageConfigurer.getFailureCollector()); // validate before macros are substituted
+    FailureCollector collector = stageConfigurer.getFailureCollector();
+    if (!config.getConnection().canAttemptToEstablishConnection()) {
+      config.validateFilters(collector);
+      return;
+    }
+    OAuthInfo oAuthInfo = null;
+    try {
+      oAuthInfo = Authenticator.getOAuthInfo(config.getConnection().getAuthenticatorCredentials());
+    } catch (Exception e) {
+      collector.addFailure("Error encountered while establishing connection: " + e.getMessage(),
+                      "Please verify authentication properties are provided correctly")
+              .withStacktrace(e.getStackTrace());
+      throw collector.getOrThrowException();
+    }
+    config.validate(stageConfigurer.getFailureCollector(), oAuthInfo); // validate before macros are substituted
 
     // if connection params are macro or null
     if (config.getConnection() != null) {
       if (!config.getConnection().canAttemptToEstablishConnection()) {
         return;
       }
     }
-    config.validateSObjects(stageConfigurer.getFailureCollector());
+    config.validateSObjects(stageConfigurer.getFailureCollector(), oAuthInfo);
     stageConfigurer.setOutputSchema(null);
   }
 
   @Override
   public void prepareRun(BatchSourceContext context) throws ConnectionException {
     FailureCollector collector = context.getFailureCollector();
-    config.validate(collector);
-    config.validateSObjects(collector);
+    OAuthInfo oAuthInfo = null;
+    try {
+      oAuthInfo = Authenticator.getOAuthInfo(config.getConnection().getAuthenticatorCredentials());
+    } catch (Exception e) {
+      collector.addFailure("Error encountered while establishing connection: " + e.getMessage(),
+                      "Please verify authentication properties are provided correctly")
+              .withStacktrace(e.getStackTrace());
+      throw collector.getOrThrowException();
+    }
+    config.validate(collector, oAuthInfo);
+    config.validateFilters(collector);
```

Comment:
can we combine all these checks? Are these not always called together?

Reply:
We have to segregate it based on whether it requires connection to do the validation or not.